src/Microsoft.VisualStudio.Validation/Requires.cs

```diff
@@ -219,8 +219,47 @@ public static void NotNullOrEmpty<T>([ValidatedNotNull, NotNull] IEnumerable<T>
                 throw new ArgumentNullException(parameterName);
             }
 
-            using IEnumerator<T> enumerator = values.GetEnumerator();
-            if (!enumerator.MoveNext())
+            bool isEmpty;
+            if (values is ICollection<T> collection)
+            {
+                isEmpty = collection.Count == 0;
+            }
+            else if (values is IReadOnlyCollection<T> readOnlyCollection)
+            {
+                isEmpty = readOnlyCollection.Count == 0;
+            }
+            else
+            {
+                using IEnumerator<T> enumerator = values.GetEnumerator();
+                isEmpty = !enumerator.MoveNext();
+            }
+
+            if (isEmpty)
+            {
+                throw new ArgumentException(Format(Strings.Argument_EmptyArray, parameterName), parameterName);
+            }
+        }
+
+        /// <summary>
+        /// Throws an exception if the specified parameter's value is null,
+        /// has no elements.
+        /// </summary>
+        /// <typeparam name="T">The type of value in the collection.</typeparam>
+        /// <param name="values">The value of the argument.</param>
+        /// <param name="parameterName">The name of the parameter to include in any thrown exception.</param>
+        /// <exception cref="ArgumentException">Thrown if the tested condition is false.</exception>
+        [DebuggerStepThrough]
+        public static void NotNullOrEmpty<T>([ValidatedNotNull, NotNull] ICollection<T> values, string? parameterName)
```

Comment:
What about the point you made in the issue:

> Unfortunately it's unclear which interfaces to support. With overloads for both `ICollection<T>` and `IReadOnlyCollection<T>`, calling `Requires.NotNullOrEmpty(new List<int>(), "")` would not compile due to ambiguity, which isn't a great experience

Why can we now add the overload?

Do we have a unit test that demonstrates that this isn't a concern any more?